cmd/snap-bootstrap/hybrid_users_import.go

```diff
@@ -0,0 +1,462 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/strutil"
+)
+
+type user struct {
+	name   string
+	uid    int
+	gid    int
+	groups []string
+
+	// these are the original lines from their respective files
+	passwdEntry string
+	shadowEntry string
+}
+
+func importHybridUserData(rootToImport, referenceRoot string) error {
+	outputDir := filepath.Join(dirs.SnapRunDir, "hybrid-users")
+	if err := os.MkdirAll(outputDir, 0755); err != nil {
+		return err
+	}
+
+	return mergeAndWriteLoginFiles(rootToImport, referenceRoot, []string{"sudo", "admin"}, outputDir)
+}
+
+func userFilter(targetGroups []string) func(user) bool {
+	return func(u user) bool {
+		// we always want to import the root user
+		if u.uid == 0 {
+			return true
+		}
+
+		// don't consider any system users or users that have a system group as
+		// their primary group
+		if u.uid < 1000 || u.gid < 1000 {
+			return false
+		}
+
+		// only consider users that are in the groups that we're specifically
+		// importing
+		for _, tg := range targetGroups {
+			if strutil.ListContains(u.groups, tg) {
+				return true
+			}
+		}
+
+		return false
+	}
+}
+
+func groupFilter(users map[string]user) func(group) bool {
+	return func(g group) bool {
+		// always import the root group
+		if g.gid == 0 {
+			return true
+		}
+
+		// in addition to importing the users that are in the given groups, we also
+		// import the primary groups that those users are in. these will most likely
+		// all be unique, but there is a chance that users share a primary group.
+		for _, u := range users {
+			if u.gid == g.gid {
+				return true
+			}
+		}
+
+		return false
+	}
+}
+
+func mergeAndWriteLoginFiles(importRoot, referenceRoot string, targetGroups []string, outputDir string) error {
+	users, err := parseUsers(importRoot, userFilter(targetGroups))
+	if err != nil {
+		return err
+	}
+
+	groups, err := parseGroups(importRoot, groupFilter(users))
+	if err != nil {
+		return err
+	}
+
+	// update the passwd and shadow files to contain the lines for the users
+	// that we're importing
+	err = mergeAndWriteUserFiles(referenceRoot, outputDir, users)
+	if err != nil {
+		return err
+	}
+
+	// update the group and gshadow files to add our users to any existing
+	// groups, and add any new primary groups that we're importing.
+	if err := mergeAndWriteGroupFiles(referenceRoot, outputDir, users, groups); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func mergeAndWriteUserFiles(originalRoot string, outputDir string, users map[string]user) error {
+	// only import users that are either root or have a uid < 1000 and gid <
+	// 1000. the base shouldn't contain anything that doesn't fit this criteria,
+	// but it is best to be safe.
+	sourceUsers, err := parseUsers(originalRoot, func(u user) bool {
+		return (u.gid == 0 && u.uid == 0) || (u.uid < 1000 && u.gid < 1000)
```

Comment:
I think we should take all that is part of the base, without filters (also, you could use just the < 1000 conditions)

Reply:
Can we guarantee that the base will never have anything with uid/gid > 1000? I know it shouldn't, but this seems safer. Otherwise we could have uid/gid conflicts. 

Alternatively, we could account for this and "fix" the uids and gids, but that would really make things complex.

Reply:
The only one with uid >= 1000 is `nobody:x:65534:65534:nobody:/nonexistent:/usr/sbin/nologin`, and classic has the same user. My fear is removing some user from the db that is needed by UC to work properly. That such a user would shadow a user from the hybrid rootfs would be strange. But neither problem should actually happen, so maybe we should make the simpler thing.

Reply:
Hrm, that is a good example. Okay, in that case we should import everything from the base, since as written we'd end up without a `nobody` user.